# # TAG ALONG CODE

In [5]:
import pandas as pd
import numpy as np
import seaborn as sns
!pip install --upgrade pandas openpyxl
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error

In [6]:
# Load the dataset
data = pd.read_csv("C:\HAMOYE\energydata_complete.csv")

In [6]:
data

date  Appliances  lights         T1       RH_1  \
0      2016-01-11 17:00:00          60      30  19.890000  47.596667   
1      2016-01-11 17:10:00          60      30  19.890000  46.693333   
2      2016-01-11 17:20:00          50      30  19.890000  46.300000   
3      2016-01-11 17:30:00          50      40  19.890000  46.066667   
4      2016-01-11 17:40:00          60      40  19.890000  46.333333   
...                    ...         ...     ...        ...        ...   
19730  2016-05-27 17:20:00         100       0  25.566667  46.560000   
19731  2016-05-27 17:30:00          90       0  25.500000  46.500000   
19732  2016-05-27 17:40:00         270      10  25.500000  46.596667   
19733  2016-05-27 17:50:00         420      10  25.500000  46.990000   
19734  2016-05-27 18:00:00         430      10  25.500000  46.600000   

              T2       RH_2         T3       RH_3         T4  ...         T9  \
0      19.200000  44.790000  19.790000  44.730000  19.000000  ...  17.033333   
1      19.200000  44.722500  19.790000  44.790000  19.000000  ...  17.066667   
2      19.200000  44.626667  19.790000  44.933333  18.926667  ...  17.000000   
3      19.200000  44.590000  19.790000  45.000000  18.890000  ...  17.000000   
4      19.200000  44.530000  19.790000  45.000000  18.890000  ...  17.000000   
...          ...        ...        ...        ...        ...  ...        ...   
19730  25.890000  42.025714  27.200000  41.163333  24.700000  ...  23.200000   
19731  25.754000  42.080000  27.133333  41.223333  24.700000  ...  23.200000   
19732  25.628571  42.768571  27.050000  41.690000  24.700000  ...  23.200000   
19733  25.414000  43.036000  26.890000  41.290000  24.700000  ...  23.200000   
19734  25.264286  42.971429  26.823333  41.156667  24.700000  ...  23.200000   

          RH_9      T_out  Press_mm_hg     RH_out  Windspeed  Visibility  \
0      45.5300   6.600000        733.5  92.000000   7.000000   63.000000   
1      45.5600   6.483333        733.6  92.000000   6.666667   59.166667   
2      45.5000   6.366667        733.7  92.000000   6.333333   55.333333   
3      45.4000   6.250000        733.8  92.000000   6.000000   51.500000   
4      45.4000   6.133333        733.9  92.000000   5.666667   47.666667   
...        ...        ...          ...        ...        ...         ...   
19730  46.7900  22.733333        755.2  55.666667   3.333333   23.666667   
19731  46.7900  22.600000        755.2  56.000000   3.500000   24.500000   
19732  46.7900  22.466667        755.2  56.333333   3.666667   25.333333   
19733  46.8175  22.333333        755.2  56.666667   3.833333   26.166667   
19734  46.8450  22.200000        755.2  57.000000   4.000000   27.000000   

       Tdewpoint        rv1        rv2  
0       5.300000  13.275433  13.275433  
1       5.200000  18.606195  18.606195  
2       5.100000  28.642668  28.642668  
3       5.000000  45.410389  45.410389  
4       4.900000  10.084097  10.084097  
...          ...        ...        ...  
19730  13.333333  43.096812  43.096812  
19731  13.300000  49.282940  49.282940  
19732  13.266667  29.199117  29.199117  
19733  13.233333   6.322784   6.322784  
19734  13.200000  34.118851  34.118851  

[19735 rows x 29 columns]

In [7]:
# View the dataset
data.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

In [8]:
# View the dataset
data.tail()

date  Appliances  lights         T1       RH_1  \
19730  2016-05-27 17:20:00         100       0  25.566667  46.560000   
19731  2016-05-27 17:30:00          90       0  25.500000  46.500000   
19732  2016-05-27 17:40:00         270      10  25.500000  46.596667   
19733  2016-05-27 17:50:00         420      10  25.500000  46.990000   
19734  2016-05-27 18:00:00         430      10  25.500000  46.600000   

              T2       RH_2         T3       RH_3    T4  ...    T9     RH_9  \
19730  25.890000  42.025714  27.200000  41.163333  24.7  ...  23.2  46.7900   
19731  25.754000  42.080000  27.133333  41.223333  24.7  ...  23.2  46.7900   
19732  25.628571  42.768571  27.050000  41.690000  24.7  ...  23.2  46.7900   
19733  25.414000  43.036000  26.890000  41.290000  24.7  ...  23.2  46.8175   
19734  25.264286  42.971429  26.823333  41.156667  24.7  ...  23.2  46.8450   

           T_out  Press_mm_hg     RH_out  Windspeed  Visibility  Tdewpoint  \
19730  22.733333        755.2  55.666667   3.333333   23.666667  13.333333   
19731  22.600000        755.2  56.000000   3.500000   24.500000  13.300000   
19732  22.466667        755.2  56.333333   3.666667   25.333333  13.266667   
19733  22.333333        755.2  56.666667   3.833333   26.166667  13.233333   
19734  22.200000        755.2  57.000000   4.000000   27.000000  13.200000   

             rv1        rv2  
19730  43.096812  43.096812  
19731  49.282940  49.282940  
19732  29.199117  29.199117  
19733   6.322784   6.322784  
19734  34.118851  34.118851  

[5 rows x 29 columns]

In [10]:
# Explore the dataset
data.shape

(19735, 29)

In [11]:
# Explore the dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

In [12]:
# Explore the dataset
data.isnull().sum()

date           0
Appliances     0
lights         0
T1             0
RH_1           0
T2             0
RH_2           0
T3             0
RH_3           0
T4             0
RH_4           0
T5             0
RH_5           0
T6             0
RH_6           0
T7             0
RH_7           0
T8             0
RH_8           0
T9             0
RH_9           0
T_out          0
Press_mm_hg    0
RH_out         0
Windspeed      0
Visibility     0
Tdewpoint      0
rv1            0
rv2            0
dtype: int64

## Question 17
### From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the Root Mean Squared error in three D.P?

In [123]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
import math
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso

In [20]:
y = data["T6"]
x = data["T2"]

In [21]:
# Split the features and target dataset into train and test sets

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [22]:
x

0        19.200000
1        19.200000
2        19.200000
3        19.200000
4        19.200000
           ...    
19730    25.890000
19731    25.754000
19732    25.628571
19733    25.414000
19734    25.264286
Name: T2, Length: 19735, dtype: float64

In [25]:
# Fit model to the training data
model = LinearRegression()

x_train_2d = x_train.values.reshape(-1, 1)
model.fit(x_train_2d, y_train)

LinearRegression()

In [27]:
# y_pred = model.predict(x_test)

x_test_2d = x_test.values.reshape(-1, 1)
y_pred = model.predict(x_test_2d)

In [30]:
x_test

8980     17.760000
2754     21.290000
9132     17.633333
14359    23.890000
8875     16.463333
           ...    
9815     20.100000
9925     20.890000
14088    23.133333
17356    22.764286
17547    26.200000
Name: T2, Length: 3947, dtype: float64

In [31]:
y_pred

array([ 2.15562488, 10.01834433,  1.87348764, ..., 14.12418366,
       13.30216726, 20.95487478])

In [32]:
# Evaluate the regression metrics for the trained model

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [33]:
print(f"MSE => {round(mse, 2)}") 
print(f"R2 => {round(r2, 2)}")

MSE => 13.2
R2 => 0.64


In [43]:
# What is the Root Mean Squared error in three D.P?
print(f"RMSE => {round (np.sqrt(mse), 3)}")

RMSE => 3.633


### Answer = 3.633

## Question 18

### Remove the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Normalize the dataset using the MinMaxScaler (Hint: Use the MinMaxScaler fit_transform and transform methods on the train and test set respectively). Run a multiple linear regression using the training set. Answer the following questions:

In [81]:
# Separate the target and features

x = data.drop(columns=["date", "lights", "Appliances"])
y = data["Appliances"]

In [82]:
# Separate the train test and test dataset
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

In [83]:
# Normalize the features dataset
scaler = MinMaxScaler()
#scaler = scaler.fit(x_train)
x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.transform(x_test)

In [84]:
# Fit model to the training data
model = LinearRegression()

model.fit(x_train_scaled, y_train)

LinearRegression()

In [85]:
y_pred = model.predict(x_test_scaled)

In [108]:
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

In [109]:
print(f"MAE => {round(mae, 3)}")
print(f"MSE => {round(mse, 3)}") 
print(f"R2 => {round(r2, 3)}")

MAE => 53.643
MSE => 8768.536
R2 => 0.149


## Question 18

In [89]:
# What is the Mean Absolute Error (in three decimal places) for the  training set?

print(f"MAE => {round(mae, 3)}")

MAE => 53.643


## Question 19

In [94]:
# What is the Root Mean Squared Error (in three decimal places) for the training set?

mse_train = mean_squared_error(y_train, y_train_pred)

In [95]:
rmse_train = mean_squared_error(y_train, y_train_pred, squared=False)

In [104]:
print(f"RMSE for the training set => {round (rmse_train, 3)}")

RMSE for the training set => 95.216


## Question 20

In [111]:
# What is the Mean Absolute Error (in three decimal places) for test set?

In [112]:
print(f"MAE for the test set => {round (mae_test, 3)}")

MAE for the test set => 53.643


## Question 21

In [106]:
# What is the Root Mean Squared Error (in three decimal places) for test set?

print(f"RMSE => {round (np.sqrt(mse), 3)}")

RMSE => 93.64


## Question 22

In [110]:
# Did the Model above overfit to the training set?

#### NO!

## Question 23

In [115]:
# The Ridge regression model

ridge_model = Ridge()

In [116]:
# Fit the model on the scaled training dataset

ridge_model.fit(x_train_scaled, y_train)

Ridge()

In [117]:
# Predict the test data 

y_pred_ridge = ridge_model.predict(x_test_scaled)

In [118]:
rmse_test_ridge = np.sqrt(mean_squared_error(y_test, y_pred_ridge))

In [120]:
print(f"RMSE for the test set with Ridge Regression => {round (rmse_test_ridge, 3)}")

RMSE for the test set with Ridge Regression => 93.709


In [121]:
# Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

##### NO significant change 

## Question 24 

In [122]:
#  How many of the features have non-zero feature weights?

In [124]:
# Lasso Regression Model

lasso_model = Lasso()

In [125]:
# Fit the model on the scaled training dataset

lasso_model.fit(x_train_scaled, y_train)

Lasso()

In [126]:
# Get the feature weights

feature_weights = lasso_model.coef_

In [128]:
# Count the number of features with non-zero weights

non_zero_weights = np.count_nonzero(feature_weights)

In [129]:
print(f"Number of features with non-zero weights => {non_zero_weights}")

Number of features with non-zero weights => 4


## Question 25 

In [130]:
# What is the new RMSE with the Lasso Regression on the test set?

In [131]:
# Predict the Lasso model output

y_pred_lasso = lasso_model.predict(x_test_scaled)

In [132]:
rmse_test_lasso = np.sqrt(mean_squared_error(y_test, y_pred_lasso))

In [134]:
print(f"RMSE for the test set with Lasso Regression => {round (rmse_test_lasso, 3)}")

RMSE for the test set with Lasso Regression => 99.424
